<div style="background: #6b7b8c; color: white; padding: 10px; font-size: 18pt; line-height: 1.5">
<b><center> Анализ базы данных (SQL)<b>
    
</div>

## Оглавление

[Описание задачи](#description)  
[Импорты и настройки](#imports)
   
[1 Изучение общей информации](#start) 
* [1.1 Таблица books](#books)
* [1.2 Таблица authors](#authors)
* [1.3 Таблица publishers](#publishers)
* [1.4 Таблица ratings](#ratings)
* [1.5 Таблица reviews](#reviews)

[2 Анализ данных](#eda)
* [2.1 Количество книг после 1 января 2000 года](#quantity)
* [2.2 Количество обзоров и средняя оценка по книгам](#review)    
* [2.3 Издательство, которое выпустило наибольшее число книг толще 50 страниц](#publisher)
* [2.4 Автор с самой высокой средней оценкой книг (книги с 50 и более оценками)](#author)
* [2.5 Среднее количество обзоров от пользователей, которые поставили больше 50 оценок](#avg_reviews)

[Общий вывод](#total_conclusion)

<a id='description'></a>
## Описание задачи

**Цель** - проанализировать базу данных купленного сервиса для чтения книг по подписке для дальнейшего формулирования ценностного предложения для нового продукта.

Задачи:
* познакомиться с хранящимися в базе данными,
* проанализировать результаты запросов.

Данные расположены в 5 таблицах, содержащих информацию о книгах, издательствах, авторах, пользовательских оценках и обзорах книг.

<a id='imports'></a>
## Импорты и настройки

In [1]:
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432, # порт подключения
 'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

<a id='start'></a>
## 1 Изучение общей информации

In [5]:
def look_table_from_sql(table):
    '''Display head and information about table in Database'''
    
    display(pd.io.sql.read_sql('''SELECT * FROM {} LIMIT 3'''.format(table), con = engine))
    display()
    display(pd.io.sql.read_sql('''SELECT * FROM {}'''.format(table), con = engine).describe())

<a id='books'></a>
### 1.1 Таблица books

In [7]:
look_table_from_sql('books')

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135

book_id    author_id   num_pages  publisher_id
count  1000.000000  1000.000000  1000.00000   1000.000000
mean    500.500000   320.417000   389.11100    171.270000
std     288.819436   181.620172   229.39014     99.082685
min       1.000000     1.000000    14.00000      1.000000
25%     250.750000   162.750000   249.00000     83.000000
50%     500.500000   316.500000   352.00000    177.500000
75%     750.250000   481.000000   453.00000    258.000000
max    1000.000000   636.000000  2690.00000    340.000000

<a id='authors'></a>
### 1.2 Таблица authors

In [8]:
look_table_from_sql('authors')

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie

author_id
count  636.000000
mean   318.500000
std    183.741666
min      1.000000
25%    159.750000
50%    318.500000
75%    477.250000
max    636.000000

<a id='publishers'></a>
### 1.3 Таблица publishers

In [9]:
look_table_from_sql('publishers')

publisher_id  publisher
0             1        Ace
1             2   Ace Book
2             3  Ace Books

publisher_id
count    340.000000
mean     170.500000
std       98.293777
min        1.000000
25%       85.750000
50%      170.500000
75%      255.250000
max      340.000000

<a id='ratings'></a>
### 1.4 Таблица ratings

In [10]:
look_table_from_sql('ratings')

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5

rating_id      book_id       rating
count  6456.000000  6456.000000  6456.000000
mean   3228.500000   510.574195     3.928284
std    1863.831001   284.141636     0.943303
min       1.000000     1.000000     1.000000
25%    1614.750000   291.000000     3.000000
50%    3228.500000   506.000000     4.000000
75%    4842.250000   750.000000     5.000000
max    6456.000000  1000.000000     5.000000

<a id='reviews'></a>
### 1.5 Таблица reviews

In [11]:
look_table_from_sql('reviews')

review_id  book_id      username  \
0          1        1  brandtandrea   
1          2        1    ryanfranco   
2          3        2      lorichen   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...

review_id      book_id
count  2793.000000  2793.000000
mean   1397.000000   504.693161
std     806.413976   288.472931
min       1.000000     1.000000
25%     699.000000   259.000000
50%    1397.000000   505.000000
75%    2095.000000   753.000000
max    2793.000000  1000.000000

<a id='eda'></a>
## 2 Анализ данных

<a id='quantity'></a>
### 2.1 Количество книг после 1 января 2000 года

In [8]:
# количество новых книг
query = ''' SELECT COUNT(book_id) AS cnt_book
            FROM books
            WHERE publication_date >= DATE('2000-01-01')
        '''

In [9]:
pd.io.sql.read_sql(query, con = engine)

cnt_book
0       821

In [10]:
# количество новых книг и их доля в базе
query = ''' SELECT 
            SUM(CASE WHEN publication_date >= DATE('2000-01-01') 
            THEN 1 ELSE 0
            END) AS cnt_book, 
            SUM(CASE WHEN publication_date >= DATE('2000-01-01') 
            THEN 1 ELSE 0
            END) * 100 / count(*) AS share_of_new_books
            FROM books
        '''

In [11]:
pd.io.sql.read_sql(query, con = engine)

cnt_book  share_of_new_books
0       821                  82

<a id='review'></a>
### 2.2 Количество обзоров и средняя оценка по книгам

In [12]:
query = ''' SELECT title, SUM(cnt_review) AS cnt_review, AVG(avg_raiting) AS avg_raiting
            FROM books 
            LEFT JOIN (SELECT book_id, COUNT(review_id) AS cnt_review 
                        FROM reviews
                        GROUP BY book_id) AS sub
            ON books.book_id = sub.book_id
            LEFT JOIN (SELECT book_id, AVG(rating) AS avg_raiting
                        FROM ratings 
                        GROUP BY book_id) AS sub_2
            ON books.book_id = sub_2.book_id
            GROUP BY books.title
        '''

In [13]:
pd.io.sql.read_sql(query, con = engine)

title  cnt_review  \
0                            The Count of Monte Cristo         5.0   
1                              Count Zero (Sprawl  #2)         2.0   
2    The Botany of Desire: A Plant's-Eye View of th...         2.0   
3                                 The Poisonwood Bible         5.0   
4                                 The Canterbury Tales         3.0   
..                                                 ...         ...   
994  In the Heart of the Sea: The Tragedy of the Wh...         3.0   
995                           Of Love and Other Demons         2.0   
996               Welcome to Temptation (Dempseys  #1)         2.0   
997                      World's End (The Sandman  #8)         2.0   
998                                  Holes (Holes  #1)         5.0   

     avg_raiting  
0       4.217391  
1       2.500000  
2       3.500000  
3       4.363636  
4       3.333333  
..           ...  
994     3.333333  
995     4.500000  
996     5.000000  
997     4.500000  
998     3.967742  

[999 rows x 3 columns]

<a id='publisher'></a>
### 2.3 Издательство, которое выпустило наибольшее число книг толще 50 страниц

In [16]:
# 1 способ
query = ''' SELECT publisher, COUNT(book_id) as cnt_book 
            FROM books LEFt JOIN publishers ON books.publisher_id = publishers.publisher_id
            WHERE num_pages > 50
            GROUP BY publisher
            ORDER BY cnt_book DESC
            LIMIT 1
        '''

In [17]:
pd.io.sql.read_sql(query, con = engine)

publisher  cnt_book
0  Penguin Books        42

In [18]:
# 2 способ - через подзапрос и оконную функцию
query = ''' SELECT publisher, cnt_book
            FROM
            (SELECT publisher, COUNT(book_id) as cnt_book, 
            MAX(COUNT(book_id)) OVER() as max_cnt_book 
            FROM books LEFT JOIN publishers ON books.publisher_id = publishers.publisher_id
            WHERE num_pages > 50
            GROUP BY publisher) AS sub
            WHERE cnt_book = max_cnt_book
        '''

In [19]:
pd.io.sql.read_sql(query, con = engine)

publisher  cnt_book
0  Penguin Books        42

<a id='author'></a>
### 2.4  Автор с самой высокой средней оценкой книг (книги с 50 и более оценками)

In [20]:
# вывод топ-5 авторов, их средний рейтинг и количество оценок
query = '''
        SELECT author, 
        ROUND(AVG(avg_raiting), 2) AS avg_raiting_by_author, SUM(cnt) AS cnt_raiting
        FROM
        (SELECT author_id, title, AVG(rating) AS avg_raiting, COUNT(rating) AS cnt
        FROM books LEFT JOIN ratings ON books.book_id = ratings.book_id 
        GROUP BY title, author_id
        HAVING COUNT(rating_id) >= 50) AS sub
        LEFT JOIN authors ON sub.author_id = authors.author_id
        GROUP BY author
        ORDER BY avg_raiting_by_author DESC
        LIMIT 5
        '''

In [21]:
pd.io.sql.read_sql(query, con = engine)

author  avg_raiting_by_author  cnt_raiting
0         J.K. Rowling/Mary GrandPré                   4.28        310.0
1  Markus Zusak/Cao Xuân Việt Khương                   4.26         53.0
2                     J.R.R. Tolkien                   4.26        162.0
3                  Louisa May Alcott                   4.19         52.0
4                      Arthur Golden                   4.11         56.0

<a id='avg_reviews'></a>
### 2.5 Среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [22]:
query = '''
        SELECT ROUND(AVG(cnt_reviews)) AS avg_reviews
        FROM
        (
        SELECT COUNT(review_id) AS cnt_reviews
        FROM
        (SELECT username 
        FROM ratings
        GROUP BY username
        HAVING COUNT(rating_id) > 50) AS sub
        LEFT JOIN reviews ON sub.username = reviews.username
        GROUP BY sub.username) AS sub_2
        '''

In [23]:
pd.io.sql.read_sql(query, con = engine)

avg_reviews
0         24.0

<a id='total_conclusion'></a>
## Общий вывод

Подключились к базе данных PostgreSQL, ознакомились с содержимым 5 таблиц: `books`, `authors`, `publishers`, `ratings`, `reviews`.

Анализ данных показал, что:
1. Количество книг, выпущенных после 1 января 2000 года, составляет **821** штуку, это 82% от всех книг в базе
2. Посчитали количество обзоров и среднюю оценку для каждой из книг (количество книг 999)
3. Издательство, которое выпустило наибольшее число книг толще 50 страниц, это **Penguin Books** с 42 книгами (исключая брошюры)
4. Автор с самой высокой средней оценкой книг - это **J.K. Rowling/Mary GrandPré** со средним рейтингом 4.28, составленным по 310 оценкам пользователей
5. В среднем количество обзоров от пользователей, которые поставили больше 50 оценок, составляет **24 штуки**